File: cf_sdk/src/main/java/com/example/cf_sdk/logic/changebanksdk/model/google/PlaceDetailSerializer.java
package com.example.cf_sdk.logic.changebanksdk.model.google;

import com.example.cf_sdk.logic.changebanksdk.parameter.google.PlaceDetailParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.member.AddAddressParameters;
import com.google.gson.annotations.SerializedName;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

/**
 *
 * Response from {@link com.example.cf_sdk.logic.changebanksdk.source.GoogleDatasource} to
 * retrieve address selected.
 */

public class PlaceDetailSerializer {

    @SerializedName("result")
    private Place result;

    @SerializedName("status")
    private String status;

    public void setPlace(Place result) {
        this.result = result;
    }

    public Place getPlace() {
        return result;
    }

    public static PlaceDetailSerializer create(Place place) {
        return new PlaceDetailSerializer(place);
    }

    public PlaceDetailSerializer(Place place) {
        setPlace(place);
    }

    public AddAddressParameters convertToAddAddressParameters(PlaceDetailParameters parameters) {

        Map<String, String> headers = parameters.getHeaders();
        String street = parameters.getAddressLine1();
        String unit =  parameters.getUnit();
        String zipCode = null;
        String city = null;
        String state = null;
        String country = null;

        for (AddressComponent component : getPlace().getAddressComponents()) {
            if (Arrays.asList(component.getTypes()).contains("locality")) {
                city = component.getLong_name();
            } else if (Arrays.asList(component.getTypes()).contains("administrative_area_level_1")) {
                state = component.getShort_name();
            } else if (Arrays.asList(component.getTypes()).contains("country")) {
                country = component.getShort_name();
            } else if (Arrays.asList(component.getTypes()).contains("postal_code")) {
                zipCode = component.getLong_name();
            }
        }

        return new AddAddressParameters(headers, street, unit, zipCode, city, state, country,"USA");
    }

    public class Geometry {
        public class Location {
            Double lat;
            Double lng;
        }

        Location location;
    }

    public static class AddressComponent {
        @SerializedName("long_name")
        String long_name;

        @SerializedName("short_name")
        String short_name;

        @SerializedName("types")
        String[] types;

        public static AddressComponent create(String[] types, String shortName, String longName) {
            return new AddressComponent(types, shortName, longName);
        }

        public AddressComponent(String[] types, String shortName, String longName) {
            this.types = types;
            this.short_name = shortName;
            this.long_name = longName;
        }

        public String getLong_name() {
            return long_name;
        }

        public String getShort_name() {
            return short_name;
        }

        public String[] getTypes() {
            return types;
        }
    }

    public class Photo {
        int height;
        int width;
        String photo_reference;
    }

    public static class Place {
        @SerializedName("place_id")
        private String mPlaceId;

        @SerializedName("name")
        private String mName;

        @SerializedName("address_components")
        private List<AddressComponent> mAddressComponents;

        @SerializedName("formatted_address")
        private String mFormattedAddress;

        @SerializedName("international_phone_number")
        private String mPhoneNumber;

        @SerializedName("photos")
        private List<Photo> mPhotos;

        @SerializedName("geometry")
        private Geometry mGeometry;

        @SerializedName("icon")
        private String mIcon;

        @SerializedName("rating")
        Double mRating;

        @SerializedName("types")
        private String[] mTypes;

        @SerializedName("url")
        private String mUrl;

        @SerializedName("vicinity")
        private String mVicinity;

        @SerializedName("website")
        private String mWebsite;

        public static Place create(String placeId,
                                   String name,
                                   List<AddressComponent> addressComponents) {
            return new Place(placeId, name, addressComponents);
        }

        public Place(String placeId,
                     String name,
                     List<AddressComponent> addressComponents) {
            mPlaceId = placeId;
            mName = name;
            mAddressComponents = addressComponents;
        }

        public String getID() {
            return mPlaceId;
        }

        public String getmName() {
            return mName;
        }

        public String getAddress() {
            return mFormattedAddress;
        }

        public List<AddressComponent> getAddressComponents(){
            return mAddressComponents;
        }

        public void setAddressComponents(List<AddressComponent> addressComponents) {
            mAddressComponents = addressComponents;
        }

        @Override
        public String toString() {
            return "Place{" +
                    "mName='" + mName + '\'' +
                    '}';
        }
    }

}

File: cf_sdk/src/main/java/com/example/cf_sdk/changebankapi/model/google/PlaceDetailSerializer.kt
package com.example.sdk_no_dagger.changebankapi.model.google

import com.example.cf_sdk.changebankapi.parameter.google.PlaceDetailParameters
import com.example.cf_sdk.changebankapi.parameter.member.AddAddressParameters
import com.google.gson.annotations.SerializedName
import java.util.Arrays

/**
 * Created by gunveernatt on 9/7/17.
 *
 * retrieve address selected.
 */
class PlaceDetailSerializer(place: Place?) {
    @SerializedName("result")
    var place: Place? = null

    @SerializedName("status")
    private val status = ""

    init {
        this.place = place
    }

    fun convertToAddAddressParameters(parameters: PlaceDetailParameters): AddAddressParameters {
        val headers = parameters.headers
        val street = parameters.addressLine1
        val unit = parameters.unit
        var zipCode: String? = null
        var city: String? = null
        var state: String? = null
        var country: String? = null
        for (component in place!!.addressComponents) {
            if (Arrays.asList(*component.types).contains("locality")) {
                city = component.long_name
            } else if (Arrays.asList<String>(*component.types)
                    .contains("administrative_area_level_1")
            ) {
                state = component.short_name
            } else if (Arrays.asList<String>(*component.types).contains("country")) {
                country = component.short_name
            } else if (Arrays.asList<String>(*component.types).contains("postal_code")) {
                zipCode = component.long_name
            }
        }
        return AddAddressParameters(headers, street, unit, zipCode, city, state, country, "USA")
    }

    inner class Geometry {
        inner class Location {
            var lat = 0.0
            var lng = 0.0
        }

        var location: Location? = null
    }

    class AddressComponent(
        @field:SerializedName("types") var types: Array<String>,
        shortName: String,
        longName: String
    ) {
        @SerializedName("long_name")
        var long_name = ""

        @SerializedName("short_name")
        var short_name = ""

        init {
            short_name = shortName
            long_name = longName
        }

        companion object {
            fun create(
                types: Array<String>,
                shortName: String,
                longName: String
            ): AddressComponent {
                return AddressComponent(types, shortName, longName)
            }
        }
    }

    inner class Photo {
        var height = 0
        var width = 0
        var photo_reference: String? = null
    }

    class Place(
        @field:SerializedName("place_id") val iD: String,
        @field:SerializedName("name") private val mName: String,
        @field:SerializedName("address_components") var addressComponents: List<AddressComponent>
    ) {

        @SerializedName("formatted_address")
        val address: String? = null

        @SerializedName("international_phone_number")
        private val mPhoneNumber: String? = null

        @SerializedName("photos")
        private val mPhotos: List<Photo>? = null

        @SerializedName("geometry")
        private val mGeometry: Geometry? = null

        @SerializedName("icon")
        private val mIcon: String? = null

        @SerializedName("rating")
        var mRating: Double? = null

        @get:SerializedName("types")
        private val mTypes: Array<String>
            get() {
                TODO()
            }

        @SerializedName("url")
        private val mUrl: String? = null

        @SerializedName("vicinity")
        private val mVicinity: String? = null

        @SerializedName("website")
        private val mWebsite: String? = null
        fun getmName(): String {
            return mName
        }

        override fun toString(): String {
            return "Place{" +
                    "mName='" + mName + '\'' +
                    '}'
        }

        companion object {
            fun create(
                placeId: String,
                name: String,
                addressComponents: List<AddressComponent>
            ): Place {
                return Place(placeId, name, addressComponents)
            }
        }
    }

    companion object {
        fun create(place: Place?): PlaceDetailSerializer {
            return PlaceDetailSerializer(place)
        }
    }
}